Automattic/jetpack
# Google Analytics: Add Support for GA4 Tags with GA4 JS

Helps with 305-gh-dotcom-manage

## Changes proposed in this Pull Request:
Currently, adding a GA4 from the Google Analytics Plugin, does not result in tracking being sent through to the Google Analytics page.

This is because the incorrect Google Analytics script is being used. GA4 uses a new script, documented here, which is backwards compatible https://developers.google.com/analytics/devguides/collection/upgrade/analyticsjs

These changes identify whether the tag is a GA4 tag, and if so, uses the GA4 JS script. Otherwise, we'll continue to use the UA JS script.

That way, tracking that was used by previous users continues to work, and tracking used my new users will work as well.

GA4 Script in source index
<img width="702" alt="Screen Shot 2020-11-12 at 7 05 14 PM" src="https://user-images.githubusercontent.com/66652282/99012158-b8d14800-251b-11eb-9198-d8ed69c8f9a9.png">

Google Analytics Metrics
<img width="779" alt="Screen Shot 2020-11-12 at 7 05 06 PM" src="https://user-images.githubusercontent.com/66652282/99012189-ca1a5480-251b-11eb-9d6c-1287bbc72766.png">

## Jetpack product discussion
N/A.

## Does this pull request change what data or activity we track or use?
No.

## Testing instructions:
For WooCommerce sites:
1. Spin up a JN website
2. Set up Jetpack
3. Choose a Jetpack plan (Daily Security)
4. Set up a new Google Analytics property with a GA4 tag for your Jetpack site
5. In `https://wordpress.com/marketing/traffic/{site_id}`,  enter your GA4 tag in the Google Analytics Tracking ID
6. Go to your Jetpack site, if you inspect the source index, you should be able to find your GA4 tag, and the GA4 script documented here: https://developers.google.com/analytics/devguides/collection/upgrade/analyticsjs
7. Go to your Google Analytics page, you should see at least one user viewing your site in realtime.
8. All functionality with the UA tags should be preserved

### Testing adding custom gtag events via filter

This implementation [enables addition of custom `gtag` events](https://developers.google.com/analytics/devguides/collection/gtagjs/events) via the filter `jetpack_gtag_universal_commands`. To test this.

* Create an `mu-plugin` on your test site - call it whatever you want. Keep it simple.
* Add the following code to your `mu-plugin` - this hooks into the filter and returns an assoc array of events for inclusion in the `<script>` tag output as custom `gtag()` events:

```php
function add_jetpack_gtag_universal_commands() {

	return array(
		array(
			'event',
			'jetpack_testing_event',
			array(
				'event_category' => 'somecat',
				'event_label'    => 'somelabel',
				'value'          => 'someval',
			),
		),
		array(
			'event',
			'another_jetpack_testing_event',
			array(
				'event_category' => 'foo',
				'event_label'    => 'bar',
				'value'          => 'baz',
			),
		),
	);
}

add_filter( 'jetpack_gtag_universal_commands', 'add_jetpack_gtag_universal_commands' );
```

* Ensure your `mu-plugin` is definitely working on your site - add some debugging output and then load the front page of your site and `View Source` (remember it only loads GA code on the front of your site not in WPAdmin).
* `View Source` and search for `dataLayer` using your browser's "Search" utility.
* You should see something akin to:

```html

<script async src='https://www.googletagmanager.com/gtag/js?id=G-XXXXXXX'></script>
<script>
    window.dataLayer = window.dataLayer || [];
    function gtag() { dataLayer.push( arguments ); }
    gtag( 'js', new Date() );
    gtag( 'config', 'G-XXXXXXX' );
    gtag( "event", "jetpack_testing_event", {"event_category":"somecat","event_label":"somelabel","value":"someval"} );
    gtag( "event", "another_jetpack_testing_event", {"event_category":"foo","event_label":"bar","value":"baz"} );
</script>

```

* Go to your Google Analytics Dashboard for the property that corresponds with the `G-XXXXX` Measurement ID (aka "Tracking ID") that you used above.
* Go to `All Events` and then create x2 events by clicking on the `Create Event` button twice.
* Name your events as per the config above - namely: `jetpack_testing_event` and `another_jetpack_testing_event`.
* Make sure you save both events.
* Back on your site, go around clicking on a few pages. Do this for about 1 minute. Do it also in Incognito mode if possible.
* Back in GA Dashboard go to `Real-time` and look for the box that reads `Event count by Event name`. 
* Wait for a bit and you should see your events show up.
* One day they will probably also show up in the `All Events` section as well but that takes longer!
* For bonus points also create some more custom events as per https://developers.google.com/analytics/devguides/collection/gtagjs/events so that we can be sure we've covered all the formats required by Google.

## Proposed changelog entry for your changes:
- Support Google Analytics 4 tags with GA4 JS script
